lib/semantics/resolve-names.cc

```diff
@@ -2557,20 +2583,25 @@ const Symbol *ResolveNamesVisitor::ResolveDataRef(const parser::DataRef &x) {
 }
 
 // base is a part-ref of a derived type; find the named component in its type.
-const Symbol *ResolveNamesVisitor::FindComponent(
-    const Symbol &base, const SourceName &component) {
-  std::optional<DeclTypeSpec> type;
-  if (auto *details{base.detailsIf<ObjectEntityDetails>()}) {
-    type = details->type();
-  } else {
+Symbol *ResolveNamesVisitor::FindComponent(
+    Symbol &base, const SourceName &component) {
+  if (!ConvertToObjectEntity(base)) {
     Say2(base.lastOccurrence(),
-        "'%s' is not an object of derived type"_err_en_US, base.name(),
-        "Declaration of '%s'"_en_US);
+        "'%s' is an invalid base for a component reference"_err_en_US,
+        base.name(), "Declaration of '%s'"_en_US);
     return nullptr;
   }
+  auto *type{base.GetType()};
   if (!type) {
     return nullptr;  // should have already reported error
   }
+  if (type->category() == DeclTypeSpec::Intrinsic &&
+      type->intrinsicTypeSpec().category() == TypeCategory::Complex) {
+    auto name{component.ToString()};
+    if (name == "re" || name == "im") {
```

Comment:
If re & im are used an a component of a non complex, and are undefined, it might be nice to say that the prefix is not a complex (if you haven't ready done that).

Reply:
The error reported for `x%re` is either that `x` is not an object of derived type or, if it is, that `re` is not a component of the type.

In the first case we're guessing what the user meant: that `x` is complex or that `x` is a derived type. The former may be more likely, the latter was easier to implement.

Reply:
In this case, it's to help me as a reader of the code.

Reply:
I don't understand this comment. What are you saying should happen?

Reply:
I'm saying that my comment should have been on the bit about asserting function or subroutine.